apps/call.go

```diff
@@ -47,211 +26,101 @@ type Call struct {
-
-// Error() makes CallResponse a valid error, for convenience
-func (cr CallResponse) Error() string {
-	if cr.Type == CallResponseTypeError {
-		return cr.ErrorText
+func (c *Call) UnmarshalJSON(data []byte) error {
+	stringValue := ""
+	err := json.Unmarshal(data, &stringValue)
+	if err == nil {
+		*c = Call{
+			Path: stringValue,
+		}
+		return nil
 	}
-	return ""
-}
 
-func CallRequestFromJSON(data []byte) (*CallRequest, error) {
-	c := CallRequest{}
-	err := json.Unmarshal(data, &c)
+	// Need a type that is just like Call, but without UnmarshalJSON
+	structValue := struct {
+		Path   string      `json:"path,omitempty"`
+		Expand *Expand     `json:"expand,omitempty"`
+		State  interface{} `json:"state,omitempty"`
+	}{}
+	err = json.Unmarshal(data, &structValue)
 	if err != nil {
-		return nil, err
+		return err
 	}
-	return &c, nil
-}
 
-func CallRequestFromJSONReader(in io.Reader) (*CallRequest, error) {
-	c := CallRequest{}
-	err := json.NewDecoder(in).Decode(&c)
-	if err != nil {
-		return nil, err
+	*c = Call{
+		Path:   structValue.Path,
+		Expand: structValue.Expand,
+		State:  structValue.State,
 	}
-	return &c, nil
+	return nil
 }
 
-func NewCall(path string) Call {
+func NewCall(path string) *Call {
 	c := Call{
 		Path: path,
 	}
-	return c
+	return &c
+}
+
+func (c *Call) WithState(state interface{}) *Call {
+	clone := Call{}
+	if c != nil {
+		clone = *c
+	}
+	clone.State = state
+	return &clone
 }
 
-func (cp *Call) WithDefault(def Call) Call {
-	if cp == nil {
+func (c *Call) WithDefault(def Call) Call {
+	if c == nil {
 		return def
 	}
-	c := *cp
+	clone := *c
 
-	if c.Path == "" {
-		c.Path = def.Path
+	if clone.Path == "" {
+		clone.Path = def.Path
 	}
-	if c.Expand == nil {
-		c.Expand = def.Expand
+	if clone.Expand == nil {
+		clone.Expand = def.Expand
 	}
-	if c.State == nil {
-		c.State = def.State
+	if clone.State == nil {
+		clone.State = def.State
 	}
-	return c
+	return clone
 }
 
-func (c *CallRequest) GetValue(name, defaultValue string) string {
-	if len(c.Values) == 0 {
-		return defaultValue
-	}
-
-	s, ok := c.Values[name].(string)
-	if ok && s != "" {
-		return s
+func (c *Call) WithLocale() *Call {
+	clone := c.Clone()
+	if clone == nil {
+		clone = &Call{}
 	}
 
-	opt, ok := c.Values[name].(map[string]interface{})
-	if ok {
-		if v, ok2 := opt["value"].(string); ok2 {
-			return v
-		}
+	if clone.Expand == nil {
+		clone.Expand = &Expand{}
 	}
 
-	return defaultValue
+	clone.Expand.Locale = ExpandAll
+	return clone
 }
 
-func (c *CallRequest) BoolValue(name string) bool {
-	if len(c.Values) == 0 {
-		return false
+func (c *Call) Clone() *Call {
+	if c == nil {
+		return nil
 	}
 
-	if b, ok := c.Values[name].(bool); ok {
-		return b
+	clone := *c
+	if clone.Expand != nil {
+		cloneExpand := *clone.Expand
+		clone.Expand = &cloneExpand
 	}
 
-	opt, ok := c.Values[name].(map[string]interface{})
-	if ok {
-		if v, ok2 := opt["value"].(bool); ok2 {
-			return v
+	// Only know how to clone map values for State.
```

Comment:
@larkox - this leaves open the possibility of a partial clone - should we call the method it PartialClone()then?